nil/internal/collate/syncer.go

```diff
@@ -69,31 +60,14 @@ func NewSyncer(cfg *SyncerConfig, db db.DB, networkManager *network.Manager) (*S
 		logger: logging.NewLogger(cfg.Name).With().
 			Stringer(logging.FieldShardId, cfg.ShardId).
 			Logger(),
-		waitForSync:   &waitForSync,
-		subs:          make(map[uint64]chan struct{}),
-		blockVerifier: signer.NewBlockVerifier(cfg.ShardId, db),
+		waitForSync: &waitForSync,
+		subs:        make(map[uint64]chan struct{}),
+		validator:   validator,
 	}, nil
 }
 
-func (s *Syncer) readLastBlock(ctx context.Context) (*types.Block, common.Hash, error) {
-	rotx, err := s.db.CreateRoTx(ctx)
-	if err != nil {
-		return nil, common.EmptyHash, err
-	}
-	defer rotx.Rollback()
-
-	block, hash, err := db.ReadLastBlock(rotx, s.config.ShardId)
-	if err != nil && !errors.Is(err, db.ErrKeyNotFound) {
-		return nil, common.EmptyHash, err
-	}
-	if err == nil {
-		return block, hash, err
-	}
-	return nil, common.EmptyHash, nil
-}
-
 func (s *Syncer) shardIsEmpty(ctx context.Context) (bool, error) {
-	block, _, err := s.readLastBlock(ctx)
+	block, _, err := s.validator.GetLastBlock(ctx)
```

Comment:
I know this has worked like this for a while, but it may be a source of inconsistency if we read such things outside of transactions. You can read the last block to be one thing initially and then at some point down the line it will be different.

Reply:
This function is called once on instance startup. Here we check that data is empty. This function expects some binary response (we have some blocks in db or not).
Below we also don't care about it because we need block.Id for log message.